### DISCUSSION TITLE
Check run and Pull Reuqest

### DISCUSSION THREAD

**Discussion by juli1** [community/community#24872] [Apps, API and Webhooks]
<p>It seems to me that when a check run is created, it referenced the based and target sha. So for each commit on a branch, we have several check runs that are being shown. However, when I create a pull request for that branch, only the results from the last check run is being surfaced and shown. If I want to show all the check runs being produced on that pull request, I need to look at each commit individually.</p>
<p>So the question is: I would like to have the ability to see per commit but I would also live to have the pull request to reflect the globality of all the check runs. That will then give a good overview to the user of all the issues that have been surfaced on all the code modified during that code review.</p>
<p>How can I achieve that?</p>
<p>Thanks!</p>

**Comment by lee-dohm** [comment]
<p>I’m not sure I understand what it is you’re trying to achieve. Let me describe a situation and see if I’m understanding correctly:</p>
<ol>
<li>Alice creates a branch</li>
<li>Alice makes three commits on the branch</li>
<li>Alice creates a pull request using that branch</li>
<li>A check run executes on the last commit in that branch which fails with three errors</li>
<li>Alice creates another commit and pushes to the branch</li>
<li>A check run executes on the latest commit and fails with two errors (one from step 4 and one new one)</li>
<li>Alice creates another commit and pushes to the branch</li>
<li>A check run executes on the latest commit and passes</li>
</ol>
<p>All of this can be seen on the pull request in the timeline view and details for each failure can be obtained by clicking the red X.</p>
<p>How would surfacing the fact that there were failures that have been fixed be better communicated? For example, on <a href="https://github.com/atom/atom/pull/19373" rel="nofollow noopener">this pull request</a>.</p>

**Comment by juli1** [comment]
<p>Let’s take the example mentioned and I will add some additional information in bold:</p>
<ol>
<li>Alice creates a branch</li>
<li>Alice makes three commits on the branch <strong>(let’s call them commit 1, commit2 and commit 3)</strong>
</li>
<li>Alice creates a pull request using that branch. <strong>When the pull request will be created, a check suite event will be sent with revision X and Y. X is the sha of commit2, Y is the sha of commit 3). Other commits are NOT included in the check_suite event.</strong>
</li>
<li>A check run executes on the last commit in that branch which fails with three errors.  <strong>If we use the information from the check_suite to annotate the commit, we annotate only the difference between commit2 and commit3. And we annotate in that case commit3.</strong>
</li>
<li>Alice creates another commit and pushes to the branch. <strong>Let’s call that commit commit 4. That creates a check_suite event on the webhook with commit3 and commit4. Previous commit are not included in the check run.</strong>
</li>
<li>A check run executes on the latest commit and fails with two errors (one from step 4 and one new one). <strong>If we rely on the check_suite event, commit from step4 (commit 3) is not included.</strong>
</li>
<li>Alice creates another commit and pushes to the branch. Let’s call is commit5. <strong>That will also trigger a check_suite even on the webhook with commit4 and commit 5.</strong>
</li>
<li>A check run executes on the latest commit and passes.</li>
</ol>
<p>So there is the issue: at each commit on a branch, a check suite is created. The check suite is only the diff between the previous change and the current one. And check_run are supposed to be added to the check suite. These check run are associated with the sha of the change. And *also* surfaced in the pull request view. In other words: the pull request contains check runs that can be related only to one commit (check run coming from the check suite) but also contains check runs from the pull request.</p>
<p>The problem is that all check runs (from check suite AND pull requests) are surfaced on the same view, giving incorrect results (these checks do not apply to the same scope of change, one taking only one commit, another takes multiple). So my question are rather:</p>
<ol>
<li>
<p><strong>How check suite and check runs are related to a pull request?</strong> According <a href="https://developer.github.com/v3/checks/suites/" rel="nofollow noopener">to the documentation</a>: "<em>A check suite is a collection of the <a href="https://developer.github.com/v3/checks/runs/" rel="nofollow noopener">check runs</a> created by a single GitHub App for a specific commit."</em> However, how do all the check runs related to a pull request? I would expect the pull request (that contains multiple commits) view to aggregate the results of all check suite (that is only for one commit)</p>
</li>
<li>
<p><strong>What event and data to use to anlyze a pull request?</strong> Based on your comment, is seems that the only useful information is the pull request information and not the check suite. So what is the purpose of the check suite and can it be useful?</p>
</li>
</ol>
<p>Thanks.</p>

**Comment by lee-dohm** [comment]
<p>I feel like you may be misunderstanding how Git works. A commit hash, or SHA, refers to a commit in the repository. A commit in the repository contains the state of the entire repository at that point in time, it  <strong>does not</strong> reference a difference between two states. Older source control systems, such as SVN and CVS worked on a diff-based system like you describe. You can still get the difference between two commits out of Git, but that is not what is stored in Git or what is represented by a commit.</p>
<p>So check suites and check runs are created for specific commits, not the difference between two commits, as stated in the documentation.</p>
<p>To answer your questions:</p>
<p><strong>How are check suites and check runs associated to a pull request?</strong></p>
<p>By strict definition, they’re not. They’re associated to single commits. Some commits are associated with pull requests and there are visualizations on the PR timeline view for those check suites.</p>
<p><strong>What event and data to use to analyze a pull request?</strong></p>
<p>The way that GitHub is designed is that the result state of the list of check suites executed against the last commit of the pull request is the state of the pull request itself. This is because when the pull request is merged, the repository will either equal the last commit of the pull request, or very closely resemble it (in the case that there have been changes on the default branch since the pull request branch was created). So the state of the check suites executed against the <em>last commit</em> is the most valid predictor of the success or failure of the pull request after merging. The state of the check suites executed against any previous commit are no longer valid predictors because any problems in those previous commits could have been fixed in the latest commit. This is why the state of the check suites of the last commit are surfaced in the GitHub Pull Request UI:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/8/85bf27f86905fff5740a50140f21b399bc1cd8e2.png" data-download-href="/uploads/short-url/j5b3kpPBvzI2WbIobOhelfIViSK.png?dl=1" title="Screen Shot 2019-07-29 at 10.48.44 AM.png"><img src="https://user-images.githubusercontent.com/1038121/181087138-701f8d0d-856f-4b08-a6c2-a85136ac8c48.png" alt="Screen Shot 2019-07-29 at 10.48.44 AM.png" data-base62-sha1="j5b3kpPBvzI2WbIobOhelfIViSK" width="472" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/8/85bf27f86905fff5740a50140f21b399bc1cd8e2_2_472x500.png, https://aws1.discourse-cdn.com/github/optimized/2X/8/85bf27f86905fff5740a50140f21b399bc1cd8e2_2_708x750.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/8/85bf27f86905fff5740a50140f21b399bc1cd8e2_2_944x1000.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/8/85bf27f86905fff5740a50140f21b399bc1cd8e2_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screen Shot 2019-07-29 at 10.48.44 AM.png</span><span class="informations">1922×2032 370 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p><a href="https://github.com/atom/atom/pull/18988">Pull Request from the above screenshot</a></p>
<p>In the example above, when <code>eff2f76</code> was the last commit (March 20), the state of the pull request was that it was not ready to be merged because one or more check suites that were associated with <code>eff2f76</code> failed. Then, on May 31, rafeca added some more commits to the pull request that resulted in <code>df3352c</code>. All of the check suites associated with <code>df3352c</code> passed, so the pull request <strong>at <code>df3352c</code></strong> is ready to be merged. Or I could go in and add some more commits that include a problem and it would go back to being not ready to merge. And so on.</p>
<p>So, to summarize:</p>
<ol>
<li>
<p>A check run is associated with a single commit and returns a status when complete</p>
</li>
<li>
<p>A check suite is a collection of check runs associated with a single GitHub App and a single commit that summarizes the status of the individual check runs</p>
</li>
<li>
<p>Multiple check suites can be executed against an individual commit</p>
</li>
<li>
<p>The results of the check suites on any commit can be used to determine what problems might exist on that commit</p>
</li>
<li>
<p>For example, perhaps you’re building a cross-platform application. One of the check suites might indicate a problem with the Windows build, whereas the macOS and Linux build check suites both pass.</p>
</li>
<li>
<p>The mergeability of a pull request can be determined based on the states of the check suites of the last commit on the pull request</p>
</li>
<li>
<p>GitHub has ways that you can indicate that certain check suites are required while others are not. For example, perhaps the build verification test check suite is required to pass whereas the check suite that calculates code coverage is not required to pass</p>
</li>
</ol>
<p>I hope that helps and let me know if you have any more questions.</p>

**Comment by juli1** [comment]
<p>Thank you for your detailed answer. I will read/process it and reply later.</p>

**Comment by xcodinglabs** [comment]
<p>First of all, thanks a lot for taking the time to make such a detailed reply.</p>
<p>I will then explain my case study so that it will be more clear. What I am doing is a code review tool: I want to make sure that the code that is shipped in a pull request is good. So the code review reviews all the code being shipped/written in a code review, surfaces all the potential problems and ultimately, gives a green/red light for shipping.</p>
<p>Therefore, I need to check that the code is good for all the commits for a pull request. Not only the latest one.</p>
<p>Let me give a simple example:</p>
<ul>
<li>
<p>Step1: Alice starts a new branch. Alice add a file (code.py) that contains errors. She commits, pushes the branch and start a pull request. The check run results in a lot of warning, it does not pass.</p>
</li>
<li>
<p>Step2: Alice commit a new file on the branch and adds a new file (code2.py) that does not contains any error. The check passes.</p>
</li>
</ul>
<p>The pull request will reflect the status of the last commit only and therefore, shows a pass (because it takes only the value of the latest commit) instead of showing the value of all the commits in the pull requests.</p>
<p>Does that explain the problem here? I think the problem is that the check for a code review is that we only want to analyze the subset of code being modified and not the whole repository (and this is what is also specified by a <code>check_suite</code> with the <code>before</code> and <code>after</code> parameters).</p>

**Comment by juli1** [comment]
<p>Note: Replied with the account that holds the application I develop :slight_smile:</p>

**Comment by lee-dohm** [comment]
<p>Yes, I understand your problem description. You want:</p>
<ol>
<li>A method to surface the results of a check on every commit because</li>
<li>For optimization purposes, you want to only analyze the code that was changed in your check</li>
<li>Also, you want the pull request UI to surface all of that information</li>
</ol>
<p>The goal of the checks system is to give feedback to the developer of the health of the codebase at the point in time of a given commit. This result can then be used to inform the developer on whether a pull request is ready to be merged or help a team enforce rules around what code is acceptable to be merged into various stages of development (signified by branches).</p>
<p>Keeping the above goal in mind, if we somehow surfaced in the pull request UI that the pull request consists of 38 “bad” commits and 42 “good” commits, does this help the developer make the decision of whether or not to press the Merge button?</p>
<p><strong>Current</strong></p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/9/90dbc4f00352804000c2976d3f3d90bf36129e44.png" data-download-href="/uploads/short-url/kFtBM2ntgGagWYbHgEZsdHfZReA.png?dl=1" title="Screen Shot 2019-07-30 at 1.38.59 PM.png"><img src="https://user-images.githubusercontent.com/1038121/181087140-1b8a5900-86bf-47ce-ba7a-5d83c5a6baee.png" alt="Screen Shot 2019-07-30 at 1.38.59 PM.png" data-base62-sha1="kFtBM2ntgGagWYbHgEZsdHfZReA" width="690" height="206" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/9/90dbc4f00352804000c2976d3f3d90bf36129e44_2_690x206.png, https://aws1.discourse-cdn.com/github/optimized/2X/9/90dbc4f00352804000c2976d3f3d90bf36129e44_2_1035x309.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/9/90dbc4f00352804000c2976d3f3d90bf36129e44_2_1380x412.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/9/90dbc4f00352804000c2976d3f3d90bf36129e44_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">Screen Shot 2019-07-30 at 1.38.59 PM.png</span><span class="informations">1892×566 59.5 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p><strong>Proposed</strong></p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/f/fe71e34eae8990e4aea62a1b004422e63969dc12.png" data-download-href="/uploads/short-url/AiVhr4oijvggxLPOrpgW15Oz29Q.png?dl=1" title="proposed.png"><img src="https://user-images.githubusercontent.com/1038121/181087141-f5159b1c-11c3-4ae4-b8fd-1603964a14b4.png" alt="proposed.png" data-base62-sha1="AiVhr4oijvggxLPOrpgW15Oz29Q" width="690" height="206" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/f/fe71e34eae8990e4aea62a1b004422e63969dc12_2_690x206.png, https://aws1.discourse-cdn.com/github/optimized/2X/f/fe71e34eae8990e4aea62a1b004422e63969dc12_2_1035x309.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/f/fe71e34eae8990e4aea62a1b004422e63969dc12_2_1380x412.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/f/fe71e34eae8990e4aea62a1b004422e63969dc12_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">proposed.png</span><span class="informations">1892×566 64.6 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Or maybe I’m misunderstanding your vision of what you want?</p>

**Comment by ThirdEyeRose** [comment]
<p>Hi <a class="mention" href="/juli1">@juli1</a>, </p>
<p>The example that you gave doesn’t seem to be correct.</p>
<blockquote>
<p>Let me give a simple example:</p>
</blockquote>
<blockquote>
<ul>
<li>Step1: Alice starts a new branch. Alice add a file (code.py) that contains errors. She commits, pushes the branch and start a pull request. The check run results in a lot of warning, it does not pass.</li>
</ul>
</blockquote>
<blockquote>
<ul>
<li>Step2: Alice commit a new file on the branch and adds a new file (code2.py) that does not contains any error. The check passes.</li>
</ul>
</blockquote>
<p>In this example, the check on Step2 will fail unless you have corrected the errors included in code.py in the first step. This is because commits are layered sequentially and don’t stand independent from one another. That’s why only the last commit in a chain is checked.</p>
<p>Hope that helps!</p>

**Comment by juli1** [comment]
<p><span class="mention">@lee-dohmyou</span> totally understand the use case and got the problem :slight_smile:</p>
<p>Many thanks for taking the time to understand the underlying problem, I can understand it is not the common way of using checks!</p>
<p>I am not sure what is the “best” solution. Right now, what I am doing is to run check runs on pull-request events: anytime I have a pull-request event, my engine runs and finds issues that have been introduced between the source and target sha. Not sure this is the best way to do it tho, that is why I started this discussion.</p>

**Comment by lee-dohm** [comment, marked as answer]
<blockquote>
<p><a class="mention" href="/juli1">@juli1</a> wrote:</p>
</blockquote>
<blockquote>
<p>I am not sure what is the “best” solution. Right now, what I am doing is to run check runs on pull-request events: anytime I have a pull-request event, my engine runs and finds issues that have been introduced between the source and target sha. Not sure this is the best way to do it tho, that is why I started this discussion.</p>
</blockquote>
<p>One weakness I can see with this approach is that your engine wouldn’t find issues that existed before the source commit. So let’s say that you enhance your engine to detect new types of problems. Your engine would detect those problems when someone introduced a new one but old ones lingering in the code base would go unnoticed.</p>
<p>Another weakness is that sometimes issues in one piece of code can be introduced by changing some other piece of code. A simple example would be that I reference a function <code>foo</code> in many places in the code but then delete the function <code>foo</code>. Every single one of those references to <code>foo</code> would be an issue even though they haven’t been changed.</p>
<p>Perhaps how you’re analyzing things aren’t subject to these or related pitfalls, though? :man_shrugging: </p>

**Comment by juli1** [comment]
<p>You are correct: with such an approach, we do not find previous issues. But that is not the goal: we want to find issues from what is sent. This is the approach taken by usual code review tools, such as phabricator (<a href="https://www.phacility.com/phabricator/" rel="nofollow noopener">https://www.phacility.com/phabricator/</a>) or reviewboard (<a href="https://www.reviewboard.org/" rel="nofollow noopener">https://www.reviewboard.org/</a>).</p>
<p>The reason is the following: somebody might start to install my GitHub app on a old, legacy repository that already has 10,000 issues. No one wants to have all these issues flagged at each code review. Instead, one wants to make sure they are not adding more issues at each commit. This is why we are only inspecting the diff.</p>
<p>So I am staying with the current solutions but might still think of integrating the check_suite event when a change is not part of a pull request (for example, when somebody commit a change on a branch without a pull request.</p>
<p>Many kudos for your help and time to clarify how check_run and check_suite work</p>
